File: stripeline/maptools.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

from typing import Any

import stripeline._maptools as _m
import numpy as np


class ConditionMatrix:
    '''Compute the inverse condition number for pixels in a map

    This class computes the inverse condition number of the pixels in a map,
    given one or more streams of samples taken from TODs. Condition numbers
    are useful to quantify how well map-makers are able to derive the I/Q/U
    components of the sky signal. This class computes the *inverse* condition
    numbers, which is the most widely used approach: condition numbers range
    from 1 (best case, perfect I/Q/U reconstruction) to infinity (worst case),
    while inverse condition numbers range from 0 (no possibility to disentangle
    I/Q/U) to 1 (best case).

    A typical usage of this class is to create an object and repeatedly call
    the :func:`ConditionMatrix.update` method with part of all the samples
    in the TOD. When all the TODs have been processed, the function
    :func:`ConditionMatrix.to_map` can be used to trigger the computation
    of the condition numbers and produce a map.
    '''

    def __init__(self, numpix: int):
        '''Create a ConditionMatrix object

        The `numpix` parameter specifies how many pixels the map should contain.
        In the case of Healpix maps, this should be the result of a call
        to healpy.nside2npix.
        '''
        self.numpix = numpix
        self.matr = np.zeros((numpix, 9), dtype='float64', order='F')

    def update(self, pixidx: Any, angle: Any):
        '''Update the condition matrix with new samples from a TOD.

        The arrays `pixidx` and `angle` must have the same number of elements.
        The first array associates each item in `angle` with a pixel in the sky.
        '''
        print('pixidx.shape =', pixidx.shape)
        print('matr.shape =', self.matr.shape)
        _m.update_condmatr(numpix=self.numpix, pixidx=pixidx,
                           angle=angle, m=self.matr)

    def to_map(self):
        '''Compute the inverse condition numbers and return them as a map.

        A pixel in the map is set to zero either if it has not been seen
        (hit count is zero), or if the components I/Q/U cannot be determined
        at all.
        '''
        seen_mask = self.matr[:, 0] > 0
        cond_map = np.zeros(self.numpix)

        # Ordered list of all the pixels which have an hit count larger than 0
        pixels = np.arange(self.matr.shape[0])[seen_mask]

        for cur_pixel in pixels:
            cond_map[cur_pixel] = 1.0 / \
                np.linalg.cond(np.reshape(self.matr[cur_pixel], (3, 3)))

        return cond_map

def nonoise_map(signal, pixidx, num_of_pixels):
    '''Convert a timeline into a map assuming no noise.

    This function estimates the map produced from ``signal`` (a vector
    containing the timeline of measurements), assuming that each sample
    in ``signal`` was looking at the sky along the direction specified
    by each element in ``pixidx`` (a vector containing the index of
    the pixels). The value of ``num_of_pixels`` is the length of the
    vector containing the map that is returned by this function.

    This function assumes that there is *no noise at all* in ``signal``.
    '''

    assert len(signal) == len(pixidx)
    assert isinstance(num_of_pixels, int)
    assert num_of_pixels > 0

    mappixels = np.zeros(num_of_pixels)
    observed = np.zeros(num_of_pixels, dtype='bool')

    for i in range(len(signal)):
        cur_pixel_pos = pixidx[i]
        if not observed[cur_pixel_pos]:
            mappixels[cur_pixel_pos] = signal[i]
            observed[cur_pixel_pos] = True

    return mappixels

def binned_map(signal, pixidx, num_of_pixels):
    '''Convert a timeline into a map assuming white noise with zero mean.

    This function estimates the map produced from ``signal`` (a vector
    containing the timeline of measurements), assuming that each sample
    in ``signal`` was looking at the sky along the direction specified
    by each element in ``pixidx`` (a vector containing the index of
    the pixels). The value of ``num_of_pixels`` is the length of the
    vector containing the map that is returned by this function.

    This function assumes that the only kind of noise in ``signal`` is
    uncorrelated noise with zero mean and symmetric probability function.

    This function returns a tuple containing the binned map and the hit map.

    The following example loads the pointing information and the signal TOD
    from a FITS file, creates a map and saves it to disk::

        from stripeline import maptools as mt
        import healpy
        from astropy.io import fits

        # Read pointings and signal from a FITS file
        with fits.open('toi.fits') as f:
            theta, phi, signal = [f[1].data.field(x)
                                  for x in ('THETA', 'PHI', 'SIGNAL')]

        NSIDE = 256
        pixidx = healpy.ang2pix(NSIDE, theta, phi)
        m, hits = mt.binned_map(signal, pixidx, healpy.nside2npix(pixidx))

        # Save both the sky map and the hit map
        healpy.write_map('map.fits', (m, hits))
    '''

    assert len(signal) == len(pixidx)
    assert isinstance(num_of_pixels, int)
    assert num_of_pixels > 0

    mappixels = np.zeros(num_of_pixels)
    hits = np.zeros(num_of_pixels, dtype='int')

    _m.binned_map(signal, pixidx, mappixels, hits)

    return mappixels, hits

File: stripeline/_maptools.f90
! Update the array of conditioning matrices "m" with new samples from
! a TOD. The "pixidx" array contains the zero-based indexes in the Healpix
! map of the polarization angles in the array "angle". Note that
! "pixidx" must be a *zero-based* array! The routine internally does the
! conversion to Fortran indexing (which is 1-based).
subroutine update_condmatr(numpix, pixidx, angle, m)
    implicit none

    integer(kind=8), intent(in) :: numpix
    integer(kind=4), dimension(:), intent(in) :: pixidx
    real(kind=8), dimension(size(pixidx)), intent(in) :: angle
    real(kind=8), dimension(numpix, 9), intent(inout) :: m

    real(kind=8) :: cos2angle
    real(kind=8) :: sin2angle
    real(kind=8) :: sincos2angle
    integer :: i
    integer(kind=4) :: pixidx1

    do i = 1, size(pixidx)
        ! Python indexing to Fortran indexing
        pixidx1 = pixidx(i) + 1

        cos2angle = cos(2.0 * angle(i))
        sin2angle = sin(2.0 * angle(i))
        sincos2angle = sin2angle * cos2angle

        m(pixidx1, 1) = m(pixidx1, 1) + 1
        m(pixidx1, 2) = m(pixidx1, 2) + cos2angle 
        m(pixidx1, 3) = m(pixidx1, 3) + sin2angle 
        m(pixidx1, 4) = m(pixidx1, 4) + cos2angle 
        m(pixidx1, 5) = m(pixidx1, 5) + cos2angle * cos2angle 
        m(pixidx1, 6) = m(pixidx1, 6) + sincos2angle 
        m(pixidx1, 7) = m(pixidx1, 7) + sin2angle 
        m(pixidx1, 8) = m(pixidx1, 8) + sincos2angle 
        m(pixidx1, 9) = m(pixidx1, 9) + sin2angle * sin2angle 
    end do
end subroutine update_condmatr

subroutine binned_map(signal, pixidx, mappixels, hits)
    implicit none

    real(kind=8), dimension(:), intent(in) :: signal
    integer(kind=8), dimension(size(signal)), intent(in) :: pixidx
    real(kind=8), dimension(:), intent(inout) :: mappixels
    integer(kind=8), dimension(size(mappixels)), intent(inout) :: hits

    integer :: i

    mappixels = 0.0
    hits = 0

    do i = 1, size(signal)
        mappixels(pixidx(i) + 1) = mappixels(pixidx(i) + 1) + signal(i)
        hits(pixidx(i) + 1) = hits(pixidx(i) + 1) + 1
    end do

    do i = 1, size(mappixels)
        if (hits(i) .gt. 0) then
            mappixels(i) = mappixels(i) / hits(i)
        end if
    end do

end subroutine binned_map